python/lsst/dax/obscore/siav2.py

```diff
@@ -0,0 +1,703 @@
+                    for t in time_interval:
+                        if not math.isfinite(t):
+                            # Timespan uses None to indicate unbounded.
+                            times.append(None)
+                        else:
+                            times.append(astropy.time.Time(float(t), scale="utc", format="mjd"))
+                    parsed_times.append(Timespan(times[0], times[1]))
+            parsed["time"] = parsed_times
+        if calib or isinstance(calib, numbers.Integral):  # 0 is allowed.
+            parsed["calib"] = frozenset(int(n) for n in ensure_iterable(calib))
+        if pol:
+            parsed["pol"] = ensure_iterable(pol)
+        if fov:
+            parsed["fov"] = [Interval.from_string(val) for val in ensure_iterable(fov)]
+        if spatres:
+            parsed["spatres"] = [Interval.from_string(val) for val in ensure_iterable(spatres)]
+        if specrp:
+            parsed["specrp"] = [Interval.from_string(val) for val in ensure_iterable(specrp)]
+        if timeres:
+            parsed["timeres"] = [Interval.from_string(val) for val in ensure_iterable(timeres)]
+        if id:
+            parsed["id"] = ensure_iterable(id)
+        if collection:
+            parsed["collection"] = ensure_iterable(collection)
+        if facility:
+            parsed["facility"] = ensure_iterable(facility)
+        if dptype:
+            parsed["dptype"] = ensure_iterable(dptype)
+        if target:
+            parsed["target"] = ensure_iterable(target)
+        if maxrec is not None:
+            parsed["maxrec"] = int(maxrec)
+        return cls.model_validate(parsed)
+
+
+class SIAv2Handler:
+    """Process SIAv2 query parameters and convert into Butler query
+    constraints.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        Butler to query.
+    config : `ExporterConfig`
+        ObsCore configuration.
+    """
+
+    def __init__(self, butler: Butler, config: ExporterConfig):
+        self.butler = butler
+        self.config = config
+        self.warnings: list[str] = []
+
+    def get_all_instruments(self) -> list[str]:
+        """Query butler for all known instruments.
+
+        Returns
+        -------
+        instruments : `list` [ `str` ]
+            All the instrument names known to this butler.
+        """
+        return [rec.name for rec in self.butler.query_dimension_records("instrument")]
+
+    def get_band_information(
+        self, instruments: list[str], band_intervals: Iterable[Interval]
+    ) -> dict[str, Any]:
+        """Read all information from butler necessary to form a band query.
+
+        Parameters
+        ----------
+        instruments : `list` [ `str` ]
+            Instruments that could be involved in the band query.
+        band_intervals : `~collections.abc.Iterable` of `Interval`
+            The band constraints.
+
+        Returns
+        -------
+        info : `dict` [ `str`, `typing.Any` ]
```

Comment:
The contents are universe-dependent. In our case we are returning a dict of filters and bands that match the query parameters. SphereX won't be returning that though since they don't have physical filters in their data model. I suppose a could define an empty `BandInfo` model and then return a subclass in this method but that wouldn't really help with the type safety problem since "filters" and "bands" would be properties of the subclass.